### ISSUE TITLE
RuntimeError: generator raised StopIteration

### ISSUE THREAD

**Issue by Redoubts** [PyCQA/pylint#4822]
### Bug description

So I've noticed another one of these StopIteration bugs when using pylint on my codebase. I don't know if I can provide a minimal example, but seeing a few similar bugs in the issue tracker, the traceback below may be enough to track it down.

For me, this reproduces reliably with astroid==2.6.6 and pylint==2.9.6, but astroid==2.6.5 pylint==2.9.6 does not repro.

### Configuration

```ini
% pip freeze | grep -E pylint\|astroid
astroid==2.6.6
pylint==2.9.6

% python -V 
Python 3.8.7
```

### Command used

```shell
pylint
```

### Pylint output

```shell
Exception on node <~~~> in file '/~~~.py'
Traceback (most recent call last):
  File "/~~~/lib/python3.8/site-packages/astroid/inference_tip.py", line 21, in _inference_tip_cached
    return iter(_cache[func, node])
KeyError: (<function infer_typing_cast at ~~~>, <Call ~~~>)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/~~~lib/python3.8/site-packages/astroid/node_classes.py", line 345, in infer
    results = tuple(self._explicit_inference(self, context, **kwargs))
  File "/~~~/lib/python3.8/site-packages/astroid/inference_tip.py", line 23, in _inference_tip_cached
    result = func(*args, **kwargs)
  File "/~~~/lib/python3.8/site-packages/astroid/brain/brain_typing.py", line 378, in infer_typing_cast
    func = next(node.func.infer(context=ctx))
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./lint", line 77, in <module>
    Run(command + groups.get("directory", list()) + py3Files)
  File "/~~~/lib/python3.8/site-packages/pylint/lint/run.py", line 384, in __init__
    linter.check(args)
  File "/~~~/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 975, in check
    self._check_files(
  File "/~~~/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1009, in _check_files
    self._check_file(get_ast, check_astroid_module, name, filepath, modname)
  File "/~~~/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1035, in _check_file
    check_astroid_module(ast_node)
  File "/~~~/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1172, in check_astroid_module
    retval = self._check_astroid_module(
  File "/~~~/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1217, in _check_astroid_module
    walker.walk(ast_node)
  File "/~~~/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
    self.walk(child)
  File "/~~~/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
    self.walk(child)
  File "/~~~/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 77, in walk
    self.walk(child)
  [Previous line repeated 2 more times]
  File "/~~~/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 74, in walk
    callback(astroid)
  File "/~~~/lib/python3.8/site-packages/pylint/checkers/logging.py", line 226, in visit_call
    result, name = is_logger_class()
  File "/~~~/lib/python3.8/site-packages/pylint/checkers/logging.py", line 208, in is_logger_class
    for inferred in node.func.infer():
  File "/~~~/lib/python3.8/site-packages/astroid/node_classes.py", line 355, in infer
    yield from self._infer(context, **kwargs)
  File "/~~~/lib/python3.8/site-packages/astroid/decorators.py", line 136, in raise_if_nothing_inferred
    yield next(generator)
  File "/~~~/lib/python3.8/site-packages/astroid/decorators.py", line 100, in wrapped
    res = next(generator)
  File "/~~~/lib/python3.8/site-packages/astroid/inference.py", line 302, in infer_attribute
    for owner in self.expr.infer(context):
RuntimeError: generator raised StopIteration
```

### Expected behavior

No crash

### Pylint version

```shell
% pylint --version                                                                   
pylint 2.9.6
astroid 2.6.6
Python 3.8.7
```

### OS / Environment

Debian 9.11

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. We desperately need #4725 for ourselves :smile:

**Comment by DanielNoord** [comment]
@Redoubts Do you have any code examples on which this occurs? Looking at the `git blame` I think I see which commits caused this regression between `2.6.5` and `2.6.6` but to resolve this a test case would be useful!

**Comment by jacobtylerwalls** [comment]
Looks like it was solved in PyCQA/astroid#1220. Adding to 2.14 milestone to decide if we can add a testcase in pylint.

**Comment by DanielNoord** [comment]
Let's close without one.
